File: frontend/src/app/features/epargne/services/ticket-recu.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { API_BASE_URL } from '../../../core/services/api.config';
import {
  TicketDuplicataRequest,
  TicketPrintRequest,
  TicketRecuResponse,
  TicketVerificationResponse
} from '../models/ticket-recu.model';

@Injectable({ providedIn: 'root' })
export class TicketRecuService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${API_BASE_URL}/tickets-recus`;

  getById(id: number): Observable<TicketRecuResponse> {
    return this.http.get<TicketRecuResponse>(`${this.baseUrl}/${id}`);
  }

  getByOperationEpargne(operationEpargneId: number): Observable<TicketRecuResponse[]> {
    return this.http.get<TicketRecuResponse[]>(`${this.baseUrl}/by-operation-epargne/${operationEpargneId}`);
  }

  getByDemandeRetrait(demandeRetraitId: number): Observable<TicketRecuResponse[]> {
    return this.http.get<TicketRecuResponse[]>(`${this.baseUrl}/by-demande-retrait/${demandeRetraitId}`);
  }

  getByMembre(membreId: number): Observable<TicketRecuResponse[]> {
    return this.http.get<TicketRecuResponse[]>(`${this.baseUrl}/by-membre/${membreId}`);
  }

  marquerImpression(id: number, request: TicketPrintRequest = {}): Observable<TicketRecuResponse> {
    return this.http.post<TicketRecuResponse>(`${this.baseUrl}/${id}/impression`, {
      format: request.format ?? 'THERMIQUE_80MM',
      marquerImprime: request.marquerImprime ?? true,
      impressionReussie: request.impressionReussie ?? true,
      commentaire: request.commentaire
    });
  }

  genererDuplicata(id: number, motif: string): Observable<TicketRecuResponse> {
    const request: TicketDuplicataRequest = { motif };
    return this.http.post<TicketRecuResponse>(`${this.baseUrl}/${id}/duplicata`, request);
  }

  getPrintableUrl(id: number, duplicata = false): string {
    return `${this.baseUrl}/${id}/print?duplicata=${duplicata}`;
  }

  getPrintableHtml(id: number, duplicata = false): Observable<string> {
    return this.http.get(`${this.baseUrl}/${id}/print?duplicata=${duplicata}`, { responseType: 'text' });
  }

  verify(codeVerification: string): Observable<TicketVerificationResponse> {
    return this.http.get<TicketVerificationResponse>(`${this.baseUrl}/verify/${encodeURIComponent(codeVerification)}`);
  }
}

File: backend/src/main/java/com/mini/credit/service/TicketRecuService.java
package com.mini.credit.service;

import com.mini.credit.dto.document.TicketDuplicataRequest;
import com.mini.credit.dto.document.TicketPrintRequest;
import com.mini.credit.dto.document.TicketRecuGenerationRequest;
import com.mini.credit.dto.document.TicketRecuResponse;
import com.mini.credit.dto.document.TicketVerificationResponse;

import java.util.List;

public interface TicketRecuService {
    TicketRecuResponse genererDepuisOperation(TicketRecuGenerationRequest request);
    TicketRecuResponse getById(Long id);
    List<TicketRecuResponse> getByOperationEpargne(Long operationEpargneId);
    List<TicketRecuResponse> getByDemandeRetrait(Long demandeRetraitId);
    List<TicketRecuResponse> getByMembre(Long membreId);
    TicketRecuResponse marquerImpression(Long id, TicketPrintRequest request);
    TicketRecuResponse genererDuplicata(Long id, TicketDuplicataRequest request);
    String getPrintableHtml(Long id, boolean duplicata);
    TicketVerificationResponse verify(String codeVerification);
    boolean canReadTicket(Long id);
    boolean canReadMembreTickets(Long membreId);
}
